File: core/campana.mjs
// Vrdct core — Campana: neutral, re-executable US-equities market status. Deterministic:
// status is a PURE function of (unix timestamp, holiday calendar) — no clock read in the core, so
// anyone re-executes it. Used by the closed-market-soundness claim-type; reusable by any surface
// that must know whether the regulated market was open at a given instant.

export const STATUS = Object.freeze({ OPEN: 'OPEN', CLOSED: 'CLOSED', HALF_DAY: 'HALF_DAY' });

// Versioned holiday calendar — the ONLY trusted datum. Pin/verify against the official NYSE calendar.
export const CALENDAR_2026 = {
  version: 2026_01,
  // The holiday table is deliberately not a timeless rule. CMLS records outside this half-open
  // range must be rejected rather than quietly classified with 2026 holidays.
  validFrom: Date.UTC(2026, 0, 1) / 1000,
  validUntil: Date.UTC(2027, 0, 1) / 1000,
  holidays: ['2026-01-01', '2026-01-19', '2026-02-16', '2026-04-03', '2026-05-25', '2026-06-19', '2026-07-03', '2026-09-07', '2026-11-26', '2026-12-25'],
  halfDays: ['2026-11-27', '2026-12-24'],
};

const REG_OPEN_MIN = 9 * 60 + 30, REG_CLOSE_MIN = 16 * 60, HALF_CLOSE_MIN = 13 * 60;

// US Eastern DST without a tz library: EDT (UTC-4) 2nd Sun Mar 02:00 → 1st Sun Nov 02:00, else EST (-5).
function nthSundayOfMonth(year, month, n) {
  const first = new Date(Date.UTC(year, month - 1, 1)).getUTCDay();
  return 1 + ((7 - first) % 7) + (n - 1) * 7;
}
export function etOffsetHours(ts) {
  const y = new Date(ts * 1000).getUTCFullYear();
  const dstStart = Date.UTC(y, 2, nthSundayOfMonth(y, 3, 2), 7, 0, 0) / 1000;
  const dstEnd = Date.UTC(y, 10, nthSundayOfMonth(y, 11, 1), 6, 0, 0) / 1000;
  return ts >= dstStart && ts < dstEnd ? -4 : -5;
}
const pad = (n) => String(n).padStart(2, '0');
function etParts(ts) {
  const off = etOffsetHours(ts);
  const w = new Date((ts + off * 3600) * 1000);
  return { off, y: w.getUTCFullYear(), m: w.getUTCMonth() + 1, d: w.getUTCDate(), wday: w.getUTCDay(), min: w.getUTCHours() * 60 + w.getUTCMinutes(), date: `${w.getUTCFullYear()}-${pad(w.getUTCMonth() + 1)}-${pad(w.getUTCDate())}` };
}
function etWallToUnix(y, m, d, minutes) {
  const guess = Date.UTC(y, m - 1, d, Math.floor(minutes / 60), minutes % 60) / 1000;
  return guess - etOffsetHours(guess) * 3600;
}
function dayKind(dateStr, wday, cal) {
  if (wday === 0 || wday === 6) return 'weekend';
  if (cal.holidays.includes(dateStr)) return 'holiday';
  if (cal.halfDays.includes(dateStr)) return 'half';
  return 'full';
}
const closeMinFor = (kind) => (kind === 'half' ? HALF_CLOSE_MIN : REG_CLOSE_MIN);

// The pure status function.
export function marketStatus(ts, cal = CALENDAR_2026) {
  const p = etParts(ts);
  const kind = dayKind(p.date, p.wday, cal);
  const isTrading = kind === 'full' || kind === 'half';
  const closeMin = closeMinFor(kind);
  let status = STATUS.CLOSED, session_open_ts = null, session_close_ts = null;
  if (isTrading && p.min >= REG_OPEN_MIN && p.min < closeMin) {
    status = kind === 'half' ? STATUS.HALF_DAY : STATUS.OPEN;
    session_open_ts = etWallToUnix(p.y, p.m, p.d, REG_OPEN_MIN);
    session_close_ts = etWallToUnix(p.y, p.m, p.d, closeMin);
  }
  let last_close_ts = null;
  for (let back = 0; back < 10 && last_close_ts === null; back++) {
    const probe = ts - back * 86400, pp = etParts(probe), k = dayKind(pp.date, pp.wday, cal);
    if (k === 'full' || k === 'half') { const c = etWallToUnix(pp.y, pp.m, pp.d, closeMinFor(k)); if (c <= ts) last_close_ts = c; }
  }
  return { market_id: 'US_EQUITIES_REGULAR', status, dateET: p.date, etOffset: p.off, dayKind: kind, session_open_ts, session_close_ts, last_close_ts, calendar_version: cal.version };
}
export function statusNow(cal = CALENDAR_2026) { return marketStatus(Math.floor(Date.now() / 1000), cal); }

File: onchain/programs/vrdct-bond/src/reexec/campana.rs
//! Campana, on-chain. A byte-for-byte port of `core/campana.mjs`: US-equities market status as a
//! PURE function of (unix timestamp, holiday calendar). No clock read, no float, no tz library —
//! so a validator re-executes exactly what an offline `node verify.mjs` re-executes.
//!
//! Parity contract with the JS core (any divergence here is a consensus bug):
//!   • civil-date math must agree with `Date.UTC` / `getUTC*` over the supported range
//!   • ET offset: EDT (-4) from 2nd Sun Mar 02:00 to 1st Sun Nov 02:00, else EST (-5)
//!   • a HALF_DAY session is NOT `OPEN` — the JS classifier counts it on the CLOSED side

/// The pinned calendar version — the only trusted datum. Mirrors `CALENDAR_2026.version` (2026_01).
pub const CAL_2026_VERSION: u32 = 202601;

/// NYSE 2026 full closures, as `y*10000 + m*100 + d`. Mirrors `CALENDAR_2026.holidays`.
const HOLIDAYS_2026: [u32; 10] = [
    20260101, 20260119, 20260216, 20260403, 20260525, 20260619, 20260703, 20260907, 20261126,
    20261225,
];
/// NYSE 2026 early closes (13:00 ET). Mirrors `CALENDAR_2026.halfDays`.
const HALF_DAYS_2026: [u32; 2] = [20261127, 20261224];

const REG_OPEN_MIN: i64 = 9 * 60 + 30;
const REG_CLOSE_MIN: i64 = 16 * 60;

/// Days since the unix epoch for a civil date (Howard Hinnant's algorithm; exact, integer-only).
fn days_from_civil(y: i64, m: i64, d: i64) -> i64 {
    let y = if m <= 2 { y - 1 } else { y };
    let era = if y >= 0 { y } else { y - 399 } / 400;
    let yoe = y - era * 400;
    let mp = (m + 9) % 12;
    let doy = (153 * mp + 2) / 5 + d - 1;
    let doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;
    era * 146097 + doe - 719468
}

/// Inverse of `days_from_civil`.
fn civil_from_days(z: i64) -> (i64, i64, i64) {
    let z = z + 719468;
    let era = if z >= 0 { z } else { z - 146096 } / 146097;
    let doe = z - era * 146097;
    let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
    let y = yoe + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = doy - (153 * mp + 2) / 5 + 1;
    let m = if mp < 10 { mp + 3 } else { mp - 9 };
    (if m <= 2 { y + 1 } else { y }, m, d)
}

/// Day-of-month of the `n`-th Sunday. Mirrors `nthSundayOfMonth`.
fn nth_sunday_of_month(y: i64, month: i64, n: i64) -> i64 {
    let first_dow = (days_from_civil(y, month, 1) + 4).rem_euclid(7); // 0 = Sunday
    1 + (7 - first_dow) % 7 + (n - 1) * 7
}

/// US Eastern offset in hours. Mirrors `etOffsetHours`.
pub fn et_offset_hours(ts: i64) -> i64 {
    let y = civil_from_days(ts.div_euclid(86400)).0;
    let dst_start = days_from_civil(y, 3, nth_sunday_of_month(y, 3, 2)) * 86400 + 7 * 3600;
    let dst_end = days_from_civil(y, 11, nth_sunday_of_month(y, 11, 1)) * 86400 + 6 * 3600;
    if ts >= dst_start && ts < dst_end {
        -4
    } else {
        -5
    }
}

/// ET wall-clock parts of a timestamp. Mirrors `etParts`.
struct EtParts {
    ymd: u32,
    wday: i64,
    minutes: i64,
}
fn et_parts(ts: i64) -> EtParts {
    let wall = ts + et_offset_hours(ts) * 3600;
    let days = wall.div_euclid(86400);
    let secs = wall.rem_euclid(86400);
    let (y, m, d) = civil_from_days(days);
    EtParts {
        ymd: (y * 10000 + m * 100 + d) as u32,
        wday: (days + 4).rem_euclid(7),
        minutes: secs / 60,
    }
}

/// Is this instant inside a REGULAR (`STATUS.OPEN`) US-equities session?
///
/// Deliberately excludes half-days: `marketStatus` returns `HALF_DAY` there and the JS classifier
/// (`st.status === STATUS.OPEN ? open++ : closed++`) counts it on the CLOSED side. Mirrored exactly.
pub fn is_regular_open(ts: i64) -> bool {
    let p = et_parts(ts);
    if p.wday == 0 || p.wday == 6 {
        return false; // weekend
    }
    if HOLIDAYS_2026.contains(&p.ymd) || HALF_DAYS_2026.contains(&p.ymd) {
        return false;
    }
    p.minutes >= REG_OPEN_MIN && p.minutes < REG_CLOSE_MIN
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn civil_roundtrip() {
        for z in -20000..20000 {
            let (y, m, d) = civil_from_days(z);
            assert_eq!(days_from_civil(y, m, d), z);
        }
    }

    #[test]
    fn dst_boundaries_2026() {
        // 2026: DST starts Sun Mar 8, ends Sun Nov 1.
        assert_eq!(nth_sunday_of_month(2026, 3, 2), 8);
        assert_eq!(nth_sunday_of_month(2026, 11, 1), 1);
        let dst_start = days_from_civil(2026, 3, 8) * 86400 + 7 * 3600;
        assert_eq!(et_offset_hours(dst_start - 1), -5);
        assert_eq!(et_offset_hours(dst_start), -4);
    }

    #[test]
    fn session_windows() {
        // Mon 2026-08-03 13:30Z == 09:30 ET (EDT) -> open; 20:00Z == 16:00 ET -> closed.
        let day = days_from_civil(2026, 8, 3) * 86400;
        assert!(!is_regular_open(day + 13 * 3600 + 29 * 60));
        assert!(is_regular_open(day + 13 * 3600 + 30 * 60));
        assert!(is_regular_open(day + 19 * 3600 + 59 * 60));
        assert!(!is_regular_open(day + 20 * 3600));
        // Sat 2026-08-01 mid-session-hours -> weekend, closed.
        let sat = days_from_civil(2026, 8, 1) * 86400;
        assert!(!is_regular_open(sat + 15 * 3600));
        // Independence Day observed 2026-07-03 -> holiday, closed.
        let hol = days_from_civil(2026, 7, 3) * 86400;
        assert!(!is_regular_open(hol + 15 * 3600));
    }
}
